src/main/controllers/ledger-service.js

```diff
@@ -1,151 +1,100 @@
+	NETWORK_ID = require('../config').chainId,
+	SELECTED_SERVER_URL = require('./web3-service').SELECTED_SERVER_URL;
+
+const errorsMap = {
+	'ledger device: unknown_error (0x6801)': 'LEDGER_IS_TIMED_OUT'
+};
 
 module.exports = function(app) {
-	let comm = null;
-	let lastHIDPath = '';
-
-	const getNetworkId = () => {
-		return CONFIG.chainId;
-	};
-
-	const getDefaultDerivationPath = () => {
-		return "44'/60'/0'/0";
-	};
-
 	const controller = function() {};
 
-	async function closeConnection() {
-		if (!comm) {
-			return;
-		}
+	function createWeb3(accountsOffset, accountsQuantity) {
+		accountsOffset = accountsOffset || 0;
+		accountsQuantity = accountsQuantity || 6;
+		const engine = new ProviderEngine();
+		const getTransport = () => Transport.create();
+		const ledger = createLedgerSubprovider(getTransport, {
+			networkId: NETWORK_ID,
+			accountsLength: accountsQuantity,
+			accountsOffset
+		});
 
-		await comm.close_async();
-		comm = null;
+		engine.addProvider(ledger);
+		engine.addProvider(new FetchSubprovider({ rpcUrl: SELECTED_SERVER_URL }));
+		engine.start();
+		return new Web3(engine);
 	}
-	async function getEth() {
-		let list = await ledgerco.comm_node.list_async();
-		let isConnected = list && list.length;
-
-		if (isConnected) {
-			let newHDPath = list[0];
-			if (comm && lastHIDPath && lastHIDPath != newHDPath) {
-				await closeConnection();
-			}
-			lastHIDPath = newHDPath;
-		} else {
-			throw Error('Device not found');
-		}
 
-		if (!comm) {
-			comm = await ledgerco.comm_node.create_async();
+	let onError = (err, reject) => {
+		if (!reject) {
+			return;
 		}
 
-		return new ledgerco.eth(comm);
-	}
-
-	function obtainPathComponentsFromDerivationPath(derivationPath) {
-		// check if derivation path follows 44'/60'/x'/n pattern
-		const regExp = /^(44'\/6[0|1]'\/\d+'?\/)(\d+)$/;
-		const matchResult = regExp.exec(derivationPath);
-		if (matchResult === null) {
-			throw new Error(
-				"To get multiple accounts your derivation path must follow pattern 44'/60|61'/x'/n "
-			);
+		if (err instanceof TimeoutError) {
+			return reject('CUSTOM_TIMEOUT');
 		}
 
-		return { basePath: matchResult[1], index: parseInt(matchResult[2], 10) };
-	}
-
-	controller.prototype.getAccounts = async function(args) {
-		args = args || {};
-
-		let indexOffset = args.start || 0;
-		let accountsNo = args.quantity || 1;
-		let eth = await getEth();
-
-		let derivationPath = args.derivationPath || getDefaultDerivationPath();
-		const pathComponents = obtainPathComponentsFromDerivationPath(derivationPath);
-
-		let addresses = {};
-		for (let i = indexOffset; i < indexOffset + accountsNo; i += 1) {
-			let path = pathComponents.basePath + (pathComponents.index + i).toString();
-			let address = await eth.getAddress_async(path);
-			addresses[path] = address.address;
+		let message = (err.message || '').toLowerCase();
+		if (message.indexOf('denied by the user') != -1) {
+			return reject('DENIED_BY_THE_USER');
```

Comment:
Just pass in the `err` object here. Why are we creating our own less-useful error codes?